### PR TITLE
New User Management

### PR DESCRIPTION
**Pull request by murrant**
image

Waiting on #8868 and #9422 some things really need tests here.

DO NOT DELETE THIS TEXT

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

Does not collide with #9313, utilizes edituser still to manage device permissions.

### CODE REVIEW

**Comment by murrant** [comment]
@TheGreatDoc nope, thanks :)

**Comment by laf** [comment]
Doesn't save when unselecting Can Modify Password.

I'm not a fan of the new user / edit user form - it looked better as it was when it was a bit more condensed and concise, just seems to be trying to take up the entire page for the sake of it now.

The old edit user / add user pages still exist.

Does anything need to be done with `./adduser.php` or `install.php` add user?

Could you make it so that deleting a user doesn't refresh the page and just removes the row. We do this in the alert-rules if you want a code example to look at.

Aside from that, all other options adding / deleting saving / updating works for me.

**Comment by murrant** [comment]
I didn't look at the old one, just put the form together, I'll take a look.  Could definitely remove without refresh.

**Comment by laf** [comment]
> Perhaps you like modal style better?

No opinion either way. Edit could work with it but add user won't as you could be on any page.

**Comment by laf** [comment]
Looks better now :)

So just the not being able to unselect Allow user to change password option. You can enable it and it saves but not disable it.

I realise the old edituser page needs to stay as the perms are done through it but I think it would be good to redirect the standard access or /edituser/ to this new page + change the link for `Change...` in the permissions page to point to the new users mgmt page.

Same question about it affecting `./adduser.php` or `install.php` stands just in case.

**Comment by murrant** [comment]
@laf I also need to add 2fa unlock and remove.

**Comment by murrant** [comment]
no affect to ./adduser.php and install.php, doesn't touch that code.

**Comment by voipmeister** [comment]
Hi @murrant , do you still need testers for these changes?

**Comment by murrant** [comment]
@voipmeister yes, but this is out of date at the moment.  I'll update it sometime.  I was waiting so I could added automated testing of the two factor auth stuff since it has several tricky scenarios.  That is almost ready.

**Comment by voipmeister** [comment]
Understood, if you mention me here I'd be happy to do some testing 🙂

**Comment by laf** [comment]
@murrant You need this testing now?

**Comment by murrant** [comment]
I think so.  I was attempting to write some dusk tests, but I'm really bad at that.

File: html/pages/edituser.inc.php [hunk 1]
```diff
@@ -11,49 +10,46 @@
...
+if (!LegacyAuth::user()->isAdmin()) {
     include 'includes/error-no-perm.inc.php';
 } else {
     if ($vars['user_id'] && !$vars['edit']) {
-        /** @var User $user */
-        $user = User::find($vars['user_id']);
-        $user_data = $user->toArray(); // for compatibility with current code
-
+        $user_data = LegacyAuth::get()->getUser($vars['user_id']);
         echo '<p><h2>'.$user_data['realname']."</h2><a href='edituser/'>Change...</a></p>";
```
Comment (laf) [inline html/pages/edituser.inc.php]: This needs to be updated to `/users`

**Comment by murrant** [comment]
Fixed all those (and some other items, check the log if curious)

Still not sure what to do with tests, I'll probably just remove them or do something simple.

**Comment by murrant** [comment]
We have basic login tests with and without 2fa, so that is better than nothing :)

**Comment by murrant** [comment]
@voipmeister ready for testing

**Comment by voipmeister** [comment]
> 
> 
> @voipmeister ready for testing

I missed this, will try to test this weekend!

**Comment by chasgames** [comment]
Nice I like this view :+1: 

Just noting two minor issues on my instance, unfortunately i have no dev environment setup yet to test any fixes

1 - LibreNMS logo changes, i have dark mode set, so every other page pulls (https://LIBRENMS/images/librenms_logo_dark.svg), but on manage users it pulls a different logo (https://LIBRENMS/images/librenms_logo_light.svg)

2 - The custom menu doesn't work for anything under https://LIBRENMS/users , i'm not sure but i wonder if LegacyAuth::get()->canManageUsers() blocks it 
https://community.librenms.org/t/how-to-add-a-custom-menu-item/3761/4?u=chas

**Comment by murrant** [comment]
1. Must be some disparity the way the code gets the logo...
2. Custom menu support isn't in the new menu system.  Could probably be added back...

**Comment by chasgames** [comment]
Ahh thanks, 2. is rather indispensable for me, hopefully for others too as currently the only way to share custom aggregations e.g https://librenms/iftype/type=customportdescription/ and other links which don't fall under any other menu.

Thanks for fixing my other code too! I wasn't sure if LegacyAuth was still a thing :)

**Comment by murrant** [comment]
@chasgames you mean this menu??
image

**Comment by chasgames** [comment]
Ahh i remember now exploring custom_descr a few years ago, i forgot it automatically adds to the menu.

I think the reason was we have around 20 different aggregations, so putting a little symbol next to it (Font Awesome icons) made it a bit more user friendly for the rest of the users so i put it in the custom menu.

I do use it for portal links to other applications like an ip manager, and to my weather animation (https://community.librenms.org/t/weathermap-histogram/1052/4?u=chas) but i suppose this could live somewhere else.

But good point i could switch back to the custom_descr in config.php if the new menu system is not supporting it

**Comment by murrant** [comment]
@chasgames https://github.com/murrant/librenms/commit/f47e5eac2baf1627977e0eae3f937ac08fd8b63e

Also, logo theme was already fixed in that branch before you reported :)

**Comment by chasgames** [comment]
Sorry to keep updating this one , just noticed another difference;

Majority of my usernames have the dot between the firstname and lastname. Creating a new user today, but dots no longer are allowed. I'm struggling to see where this validator is in the code though.

image

**Comment by murrant** [comment]
app/Http/Requests/StoreUserRequest.php

There are better spots than this.  Often I miss comments on closed PRs ;)

**Comment by murrant** [comment]
This pull request has been mentioned on **LibreNMS Community**. There might be relevant details there:

https://community.librenms.org/t/dots-in-username/12887/2

**Pull request merged**